### ISSUE TITLE
How to implement fast non-maximal suppression in GPU?

### ISSUE THREAD

**Issue by experiencor** [tensorflow/tfjs-core#98]
Thank you for a great library.

I am trying to implement YOLO in deeplearn.js. I found that inference time is very fast (around 5ms) but the bottleneck is non-maximal suppression (around 350ms). My current code for non-maximal suppression in shown below:

```
    for (var row = 0; row < 13; row++) {
      for (var col = 0; col < 13; col++) {
        for (var box = 0; box < 5; box++) {
          const l_bound = netout.locToIndex([row, col, box, 0])
         
          var startTime = performance.now();
          // get and adjust the confidence of the box
          var confidence = this.sigmoid(netout.getValues()[l_bound+4])

          // get and adjust class likelihoods
          var classes = netout.getValues().slice(l_bound+BOX, l_bound+BOX+CLASS)
          classes = this.sofmax(classes).map((x) => (x * confidence)).map((x) => x > this.THRESHOLD ? x : 0)
          var endTime = performance.now();

          console.log(endTime - startTime)

          if (classes.reduce(( a, b ) => a + b, 0) > 0) {
            var boundBox = new BoundingBox();

            boundBox.c = confidence
            boundBox.probs = classes 

            boundBox.x = netout.getValues()[l_bound]
            boundBox.y = netout.getValues()[l_bound+1]
            boundBox.w = netout.getValues()[l_bound+2]
            boundBox.h = netout.getValues()[l_bound+3]      
            
            boundBox.x = (col + this.sigmoid(boundBox.x)) / GRID_W
            boundBox.y = (row + this.sigmoid(boundBox.y)) / GRID_H
            boundBox.w = this.ANCHORS[2 * box + 0] * Math.exp(boundBox.w) / GRID_W
            boundBox.h = this.ANCHORS[2 * box + 1] * Math.exp(boundBox.h) / GRID_H              

            boxes.push(boundBox)
          }
        }
      }
    }
```

At the moment, I have to loop through 13 * 13 * 5 = 845 boxes and carry out the same operations to find out the boxes with sufficient confidence.

Can you give me some hints on how to implement these for loops in GPU or any other ways to cut down the computation time?

**Comment by nsthorat** [comment]
Sure!

Can you give me a little more information. What is the shape of netout, and what does it represent?

Calling getValues() is expensive. The first thing I would try is caching the output of getValues() and saving it to a local variable.

Are you implementing your own softmax / sigmoid methods?

Can you link me to the entirety of your code?

Thanks :)

**Comment by experiencor** [comment]
Thanks very much for the suggestion.

The shape of my netout is (13, 13, 5, 25). It represents:

netout[:,:,:,:2] are x and y, I want to do netout[:,:,:,:2] = sigmoid(netout[:,:,:,:2])
netout[:,:,:,2:4] are w and h, I want to do netout[:,:,:,2:4] = exp(netout[:,:,:,2:4])
netout[:,:,:,5] is confidence, I want to do netout[:,:,:,5] = sigmoid(netout[:,:,:,5])
netout[:,:,:,5:] are class probabilities, I want to do netout[:,:,:,5:] = sofmax(netout[:,:,:,5:], axis=3)

after that:

zero out small probabilities in netout[:,:,:,5:] using a threshold
compute netout[:,:,:,5] * netout[:,:,:,5:] 

then choose any box (determined by first 3 axes of netour) which has sum(netout[box,5] * netout[box,5:]) > a threshold.

At the moment, I have to loop through the first 3 axes of netout to carry out these operations.

My fork of deeplearn.js is here: [https://github.com/experiencor/deeplearnjs/
](https://github.com/experiencor/deeplearnjs/)

The yolo network is here: [https://github.com/experiencor/deeplearnjs/blob/master/demos/models/yolo.ts](https://github.com/experiencor/deeplearnjs/blob/master/demos/models/yolo.ts)

The frontend code is here: [https://github.com/experiencor/deeplearnjs/blob/master/demos/yolo/yolo-demo.ts](https://github.com/experiencor/deeplearnjs/blob/master/demos/yolo/yolo-demo.ts)

**Comment by nsthorat** [comment]
Okay so a few comments.

I think the time is actually not in post-processing. There are lots of texture leaks in your model. If you call getValues() right after the infer() call, youll see a real timing.

1)
https://github.com/experiencor/deeplearnjs/blob/master/demos/models/yolo.ts#L93
See how you call "keep" everywhere - don't do that. You shouldn't have to call keep anywhere in this model (the only reason I do that in the webcam demo is because I want to render the activations).

2)
https://github.com/experiencor/deeplearnjs/blob/master/demos/models/yolo.ts#L221

These NDArrays are allocated every time this method is called, you should construct a ONE and an ALPHA cache that only gets created a single time.

3)
https://github.com/experiencor/deeplearnjs/blob/master/demos/models/yolo.ts#L138

Cache getValues(), only index one time.

4) Constructing new NDArrays during inference and uploading them causes a stall and is relatively expensive:
https://github.com/experiencor/deeplearnjs/blob/master/demos/models/yolo.ts#L102

Why exactly do you need this?

They're also untracked (when using new) so this is another texture leak (wrap them in math.track)

5)
getBoundingClientRect is super slow
https://github.com/experiencor/deeplearnjs/blob/master/demos/yolo/yolo-demo.ts#L160

Cache these numbers, don't do it every frame.

6) 
You can do a lot of the softmax, sigmoid calculations on the GPU. Check out math.slice to take a slice of an NDArray, and math.softmax and math.sigmoid.

7)
To find texture leaks, set a breakpoint here in your demo (this will be automated soon)
https://github.com/PAIR-code/deeplearnjs/blob/master/src/math/webgl/texture_manager.ts#L46

And general comments so that this will compile, you should use watch-demo (lots of missing semicolons, etc that the typescript compiler doesn't like).

**Comment by nsthorat** [comment]
A couple other things. The size of netout is quite big, if you can reduce the size by math operations on the GPU to exactly what you need for rendering before downloading to the CPU, you should get wins.

**Comment by nsthorat** [comment]
Oh, and if you sync your client to master, we can debug the speed of individual math calls (enableDebugMode on math)

**Comment by experiencor** [comment]
I discover that most of the post-processing is spent on downloading the big array netout from GPU to CPU (97%). So I have to somehow reduce the size of netout on GPU first before downloading it to CPU. This operation can be simplified to the following operation: given a big array A with a few positive elementso in GPU, how to reduce it to the smaller array B also in GPU, which contains all the positive elements of A?

Can you suggest me a way to do this simplified operation?

**Comment by nsthorat** [comment]
You should first make the changes I mentioned above, it will significantly increase performance as you have many texture leaks that are going to cause issues.

I still don't totally understand what this code is doing, or what netout represents.

Your shape is:  `[GRID_H, GRID_W, BOX, CLASS] = (13, 13, 5, 25)`

I assume this means a 13x13 coarse grid over the input space. What do each of the 5 box values mean? For a given cell, what does x, y, w, and h represent if the cell is already given? I don't see these details in the paper, is there code somewhere in python you can link me to?

Is there an upper bound for the number of boxes you can display?

**Comment by experiencor** [comment]
I have made all the changes except 6, 7 but it's still very slow. So I commented out everything in `interpret_netout` but a single call to `const values = netout.getValues()`. It takes 360ms+ for this call to getValues() to complete.

```
  interpret_netout(netout: Array4D): Array<BoundingBox> { 
    const values = netout.getValues();
```

The semantic of netout is that each cell of the 13x13 grid has 5 chances to make predictions on the bounding boxes of the objects, and so 5 boxes. Each bounding box is represented by 25 parameters: the first 4 parameters for x, y, w, h, the 5th for confidence and the last 20 parameters for 20 object classes (cat, dog, car, ...). So the CLASS is 25, which includes everything from x, y, ..., all class parameters.

You can learn more details in Section 2. Unified Dection (in https://arxiv.org/pdf/1506.02640.pdf).

There are 13*13*5 altogether and I'm only interested in boxes, whose confidences are larger than a threshold. Usually, only a few boxes satisfy this condition.

**Comment by nsthorat** [comment]
Can you commit your new changes? I don't see them.

Also, can you use vscode so that this code compiles with watch-demo? 

One trick you can do is call math.enableDebugMode(). This will throw exceptions when there are NaNs (you have a NaN), and it will also give you profiling for each individual call to see where you are spending time.

Will keep looking at this in 30 mins.

**Comment by experiencor** [comment]
@nsthorat Very sorry for the delay.

I have committed my code. In this new code, I commented out everything in `interpret_netout` except the function to download `netout` from GPU to CPU.

I am unfamiliar with vscode. I use Sublime to change the typescript code directly. But I normally run watch-demo to debug. I think you should be able too.

Regard this.math.enableDebugMode(), my this.math doesn't have enableDebugMode function. It may be because of the fact that I don't use the latest version of deeplearn.js. However, please let me know if I have to use the latest version. I will be happy to make any changes to make your debugging convenient.

**Comment by nsthorat** [comment]
Yeah you don't have the latest version of deeplearn.js, can you merge? debugMode will help you a lot, and NaNs could be causing problems.

Let me know once you merge and I'll try to run your code again!

**Comment by experiencor** [comment]
I have fixed the NaN error and merged with the latest deeplearnjs code. I found out that with this.math.enableDebugMode(), inference time increases to 2.5s+ but `netout.getValues()` takes 0.01ms. But without this.math.enableDebugMode(), `netout.getValues()` stil takes 300ms+.

**Comment by kaihuchen** [comment]
@experiencor    I am very interested in seeing YOLO working in deeplearn.js. I wonder if your code is ready for others to play with？

**Comment by experiencor** [comment]
Yes, it's ready to play with. But it's pretty sluggish. I think that the bottleneck is not actually the nonmaximal suppression processing. The original feature extraction layers of YOLO is rather expensive. I'm considering to switch to MobileNet to see if I can have a smoother detection experience.

**Comment by experiencor** [comment]
@kaihuchen Check out my PR (https://github.com/PAIR-code/deeplearnjs/pull/361), which run YOLO on MobileNet backend. A demo is accessible at https://experiencor.github.io/yolo_demo/demo.html.

**Comment by QuantumInformation** [comment]
@nsthorat when can we use NDArrayMath.enableDebugMode? I don't see it in the NDArrayMath in math.ts

**Comment by nsthorat** [comment]
Hi Nikos,

Things changed a little bit. You can enable debug mode two ways right now (we'll add a better top level method soon).

```
dl.ENV.set('DEBUG', true);
```

Or alternatively via the URL:

&dljsflags=DEBUG:true
